daffodil-runtime1-unparser/src/main/scala/org/apache/daffodil/processors/unparsers/ZeroLengthDetector.scala

```diff
@@ -83,12 +151,34 @@ sealed trait StringZeroLengthMixin { self: ZeroLengthDetector =>
 class StringZeroLengthDetector
   extends ZeroLengthDetector
   with StringZeroLengthMixin {
-  override def isZeroLength(diElement: DIElement) = isStringZeroLength(diElement)
+  override def isKnownNonZeroLength(diElement: DIElement) = isStringKnownNonZeroLength(diElement)
+  override def isKnownZeroLength(diElement: DIElement) = isStringKnownZeroLength(diElement)
 
 }
 
 sealed trait HexBinaryZeroLengthMixin { self: ZeroLengthDetector =>
-  protected def isHexBinaryZeroLength(diElement: DIElement): Boolean = {
+
+  protected def isHexBinaryKnownNonZeroLength(diElement: DIElement): Boolean = {
+    val opt = diElement.erd.optPrimType
+    Assert.invariant(opt.isDefined & (opt.get eq NodeInfo.HexBinary))
+    val result =
+      if (diElement.isNilled)
+        true // must not be zero length nil. If it was zero length nil, we wouldn't be at this code.
```

Comment:
Would it be worth adding an Assert.invariant() here to check for a zero length nil?

Reply:
Yes. I can pass hasNilZL as an argument, and test it here.

Reply:
Actually, this code gets reused in contexts where hasNilZL is true... it's just not called until we've verified that the value isn't a nilled element. So I will fix the comment, which is misleading.